x-pack/plugins/ml/public/settings/calendar/edit/new_calendar.js

```diff
@@ -0,0 +1,310 @@
+import { NewEventModal } from './new_event_modal';
+import { ImportModal } from './import_modal';
+import { ml } from 'plugins/ml/services/ml_api_service';
+import { toastNotifications } from 'ui/notify';
+
+export class NewCalendar extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isNewEventModalVisible: false,
+      isImportModalVisible: false,
+      isNewCalendarIdValid: false,
+      loading: true,
+      jobIds: [],
+      jobIdOptions: [],
+      groupIds: [],
+      groupIdOptions: [],
+      calendars: [],
+      formCalendarId: '',
+      description: '',
+      selectedJobOptions: [],
+      selectedGroupOptions: [],
+      events: [],
+      saving: false,
+      selectedCalendar: undefined,
+    };
+  }
+
+  componentDidMount() {
+    this.formSetup();
+  }
+
+  async formSetup() {
+    try {
+      const { jobIds, groupIds, calendars } = await getCalendarSettingsData();
+
+      const jobIdOptions = jobIds.map((jobId) => ({ label: jobId }));
+      const groupIdOptions = groupIds.map((groupId) => ({ label: groupId }));
+
+      const selectedJobOptions = [];
+      const selectedGroupOptions = [];
+      let eventsList = [];
+      let selectedCalendar;
+      let formCalendarId = '';
+
+      // Editing existing calendar.
+      if (this.props.calendarId !== undefined) {
+        selectedCalendar = calendars.find((cal) => cal.calendar_id === this.props.calendarId);
+
+        if (selectedCalendar) {
+          formCalendarId = selectedCalendar.calendar_id;
+          eventsList = selectedCalendar.events;
+
+          selectedCalendar.job_ids.forEach(id => {
+            if (jobIds.find((jobId) => jobId === id)) {
+              selectedJobOptions.push({ label: id });
+            } else if (groupIds.find((groupId) => groupId === id)) {
+              selectedGroupOptions.push({ label: id });
+            }
+          });
+        }
+      }
+
+      this.setState({
+        events: eventsList,
+        formCalendarId,
+        jobIds,
+        jobIdOptions,
+        groupIds,
+        groupIdOptions,
+        calendars,
+        loading: false,
+        selectedJobOptions,
+        selectedGroupOptions,
+        selectedCalendar
+      });
+    } catch (error) {
+      console.log(error);
+      this.setState({ loading: false });
+      toastNotifications.addDanger('An error occurred loading calendar form data. Try refreshing the page.');
+    }
+  }
+
+  onCreate = async () => {
+    const calendar = this.setUpCalendarForApi();
+    this.setState({ saving: true });
+
+    try {
+      await ml.addCalendar(calendar);
+      window.location = `${chrome.getBasePath()}/app/ml#/settings/calendars_list`;
+    } catch (error) {
+      console.log('Error saving calendar', error);
+      this.setState({ saving: false });
+      toastNotifications.addDanger(`An error occurred creating calendar: ${calendar.calendarId}`);
+    }
+  }
+
+  onEdit = async () => {
```

Comment:
these two functions are almost identical.
could they combined with an `isEdit` parameter or something like that?

Reply:
I kept these separate because I like keeping it clear whether we're editing or creating. Also, if we needed to pass a flag we'd have to wrap the handler in an arrow function so we could send along the flag. I prefer to avoid that since that function gets re-created every time we render.

I'm open to changing it if the team feels it's best to combine, though. 😄 
cc @walterra, @peteharverson what do you both think?

Reply:
I'm happy for these two functions to remain separate!